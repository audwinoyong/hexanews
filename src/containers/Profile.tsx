import * as React from 'react'
import Grid from '@material-ui/core/Grid'
import Divider from '@material-ui/core/Divider'
import { History } from 'history'

import Avatar from '@material-ui/core/Avatar'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import CardHeader from '@material-ui/core/CardHeader'
import ProfileCard from '../components/ProfileCard'
import Typography from '@material-ui/core/Typography'

import avatarPlaceholder from '../assets/avatar_placeholder.png'

type Props = {
  history: History
}

const styles = {
  avatar: {
    width: 50,
    height: 50,
    backgroundColor: 'green',
    alignSelf: 'center'
  },
  container: {
    marginLeft: '25%',
    marginRight: '25%',
    marginTop: '4%',
    paddingBottom: '5%'
  },
  profileContainer: {
    display: 'flex',
    marginBottom: '3%',
  },
  card: {
    boxShadow: '0 1px 4px rgba(0, 0, 0, 0.2)',
    marginTop: 36,
  }
}

class Profile extends React.Component<Props> {

  public render() {
    return (
      <div style={styles.container}>
        <ProfileCard
          name='Hillary Clinton'
          description='A politician, writer and philanthropist.'
          image={avatarPlaceholder}
        />
        <Divider style={{ marginTop: 20 }} />
        <Grid container spacing={24}>
          <Grid item xs={12} style={{ marginRight: 24 }}>
            <Card style={styles.card}>
              <CardHeader
                avatar={<Avatar style={styles.avatar}>HC</Avatar>}
                title='Hillary Clinton'
                subheader='20 August 2018'
              />
              <CardContent>
                <Typography gutterBottom variant='headline' component='h2'>
                  {'Implement Google Analytics for React Native'}
                </Typography>
                <Typography component='p'>
                  {'For almost every product that is built or launched in the market, the business and development teams testing there'}
                </Typography>
              </CardContent>
            </Card>
            <Card style={styles.card}>
              <CardHeader
                avatar={<Avatar style={styles.avatar}>HC</Avatar>}
                title='Hillary Clinton'
                subheader='11 July 2018'
              />
              <CardContent>
                <Typography gutterBottom variant='headline' component='h2'>
                  {'One Book at The Time —A Short Story'}
                </Typography>
                <Typography component='p'>
                  {'There had been times she had almost given up. All the lying, the sneaking around. But today she once more managed to continue'}
                </Typography>
              </CardContent>
            </Card>
            <Card style={styles.card}>
              <CardHeader
                avatar={<Avatar style={styles.avatar}>HC</Avatar>}
                title='Hillary Clinton'
                subheader='8 June 2018'
              />
              <CardContent>
                <Typography gutterBottom variant='headline' component='h2'>
                  {'Adjustable sidebar using Angular'}
                </Typography>
                <Typography component='p'>
                  {'Also for this tutorial I din’t put accent on CSS, I assume that you already created a sidebar which you need to make draggable.'}
                </Typography>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      </div >
    )
  }
}

export default Profile
