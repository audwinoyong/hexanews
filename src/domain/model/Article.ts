/**
 * Model for the article.
 */
export type Article = {
  title: string,
  description: string
  category?: string
  authorId?: string
}