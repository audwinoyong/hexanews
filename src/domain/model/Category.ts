/**
 * Model for category object in our application.
 */

export type Category = {
  _id?: string,
  title: string,
  description: string,
}