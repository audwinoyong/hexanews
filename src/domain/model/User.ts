/**
 * Model for the user.
 */
export type User = {
  _id: string,
  email: string,
  name: string,
  description: string,
  createdAt: string,
}