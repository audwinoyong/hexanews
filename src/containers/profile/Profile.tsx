/**
 * Profile container contains all description of current user
 */

import * as React from 'react'
import * as moment from 'moment'
import Avatar from '@material-ui/core/Avatar'
import Card from '@material-ui/core/Card'
import Button from '@material-ui/core/Button'
import CardContent from '@material-ui/core/CardContent'
import CardHeader from '@material-ui/core/CardHeader'
import Divider from '@material-ui/core/Divider'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import { CardMedia } from '@material-ui/core'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import { push } from 'connected-react-router'
import { translate } from 'react-i18next'
import { map, startCase, head } from 'lodash'

import placeholder from '../../assets/placeholder.png'
import avatarPlaceholder from '../../assets/avatar_placeholder.png'
import ProfileCard from '../../components/profileCard/ProfileCard'
import styles from './styles'
import selector, { StateProps } from './selector'
import { Article } from '../../domain/model/Article'
import { User } from '../../domain/model/User'
import Utils from '../../utils'
import { DATE_FORMAT, HEADER_LINK } from '../../constants/config'
import i18n from '../../i18n'

type Props = {
  user: User,
  dispatch: Dispatch<any>
} & StateProps

export class Profile extends React.Component<Props> {

  renderAvatar = (authorName: string) => {
    const initials = head(startCase(authorName))
    return <Avatar style={styles.avatar}>{initials}</Avatar>
  }

  renderUserArticles = () => {
    const { userArticles, dispatch, user } = this.props
    return map(userArticles, (article: Article) => {
      const sanitizedDescription = article.description.replace(/<(?:.|\n)*?>/gm, '')
      const articleImage = Utils.getFeaturedImage(article) ? Utils.getFeaturedImage(article) : placeholder
      const articleDescription = sanitizedDescription.length > 128 ? `${sanitizedDescription.substring(0, 128)}...` : sanitizedDescription
      const articleId = article._id ? article._id : ''
      return (
        <Card
          style={styles.card}
          key={article._id}
          onClick={() => dispatch(push(HEADER_LINK.articleDetail(articleId)))}
          id={`card-${article._id}`}
        >
          <CardHeader
            avatar={this.renderAvatar(user.name)}
            title={user.name}
            subheader={moment(article.createdAt).format(DATE_FORMAT)}
          />
          <CardMedia
            style={styles.media}
            image={articleImage}
          />
          <CardContent>
            <Typography gutterBottom variant='headline' component='h2'>
              {article.title}
            </Typography>
            <Typography component='p'>
              {articleDescription}
            </Typography>
          </CardContent>
        </Card>
      )
    })
  }

  public render() {
    const { dispatch, user } = this.props
    return (
      <Grid container style={styles.container as any}>
        <Grid container>
          <Grid item md={3} xs={1} />
          <Grid item md={6} xs={10} style={{ flexDirection: 'column' }}>
            <ProfileCard
              name={user.name}
              description={user.description ? user.description : ''}
              image={avatarPlaceholder}
              createdAt={user.createdAt ? user.createdAt : ''}
            />
            <Button
              id='accountButton'
              variant='outlined'
              size='small'
              component='button'
              style={styles.button}
              onClick={() => dispatch(push('/account'))}
            >
              {i18n.t('profile.myAccount')}
            </Button>
            <Divider style={styles.profileDivider} />
          </Grid>
          <Grid item md={3} xs={1} />
        </Grid>
        <Grid container spacing={24}>
          <Grid item md={3} xs={1}/>
          <Grid item md={6} xs={10} style={styles.articleContainer}>
            {this.renderUserArticles()}
          </Grid>
          <Grid item md={3} xs={1} />
        </Grid>
      </Grid >
    )
  }
}

const ConnectedProfile = connect(selector)(Profile)

export default translate('translations')(ConnectedProfile)
