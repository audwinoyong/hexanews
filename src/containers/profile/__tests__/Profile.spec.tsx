import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { omit } from 'lodash'

import { Profile } from '../Profile'
import { Article } from '../../../domain/model/Article'

Enzyme.configure({ adapter: new Adapter() })

const mockArticles: Article[] = [
  {
    _id: '1',
    title: 'abc',
    description: 'test',
    category: 'test',
    authorId: 'test',
  },
  {
    _id: '2',
    title: 'def',
    description: 'tester',
    category: 'tester',
    authorId: 'tester',
  }
]

const mockUser = {
  _id: '1',
  createdAt: 'created',
  description: 'desc',
  email: 'user@email.com',
  name: 'user'
}

describe('Profile', () => {
  it('renders Profile', () => {
    const component = renderer.create(
      <Profile
        user={mockUser}
        userArticles={mockArticles}
        dispatch={jest.fn()}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Profile without description', () => {
    const missingDescriptionUser = omit(mockUser, ['description'])
    const component = renderer.create(
      <Profile
        user={missingDescriptionUser}
        userArticles={mockArticles}
        dispatch={jest.fn()}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle button go to account page', () => {
    const dispatch = jest.fn()
    const component = (
      <Profile
        user={mockUser}
        userArticles={mockArticles}
        dispatch={dispatch}
      />
    )
    const layout = shallow(component)
    layout.find('#accountButton').first().simulate('click')
    expect(dispatch).toHaveBeenCalled()
  })

  it('handle card click go to article detail page', () => {
    const dispatch = jest.fn()
    const component = (
      <Profile
        user={mockUser}
        userArticles={mockArticles}
        dispatch={dispatch}
      />
    )
    const layout = shallow(component)
    layout.find('#card-1').first().simulate('click')
    expect(dispatch).toHaveBeenCalled()
  })
})