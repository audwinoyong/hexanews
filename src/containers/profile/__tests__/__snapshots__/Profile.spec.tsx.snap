// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders Profile 1`] = `
<div
  className="MuiGrid-container-1"
  style={
    Object {
      "marginTop": "2%",
      "paddingBottom": "5%",
    }
  }
>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
      style={
        Object {
          "flexDirection": "column",
        }
      }
    >
      <div>
        <div
          className="MuiPaper-root-99 MuiPaper-elevation1-102 MuiPaper-rounded-100 MuiCard-root-98"
          style={
            Object {
              "boxShadow": "0px 0px 0px 0px rgba(0, 0, 0, 0.2)",
              "display": "flex",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "flex": 4,
                "flexDirection": "column",
              }
            }
          >
            <div
              className="MuiCardContent-root-126"
              style={
                Object {
                  "flex": "1 0 auto",
                }
              }
            >
              <h1
                className="MuiTypography-root-127 MuiTypography-display1-131"
              >
                user
              </h1>
              <p
                className="MuiTypography-root-127 MuiTypography-body1-136"
                style={
                  Object {
                    "marginTop": 4,
                  }
                }
              >
                desc
              </p>
              <p
                className="MuiTypography-root-127 MuiTypography-body1-136"
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginTop": 12,
                  }
                }
              >
                Hexanews member since Invalid date
              </p>
            </div>
          </div>
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <img
              alt="avatar"
              src={undefined}
              style={
                Object {
                  "borderRadius": "50%",
                  "height": 100,
                  "width": 100,
                }
              }
            />
          </div>
        </div>
      </div>
      <button
        className="MuiButtonBase-root-178 MuiButton-root-152 MuiButton-text-154 MuiButton-outlined-160 MuiButton-sizeSmall-175"
        disabled={false}
        id="accountButton"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "marginLeft": 20,
          }
        }
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-153"
        >
          My Account
        </span>
        <span
          className="MuiTouchRipple-root-181"
        />
      </button>
      <hr
        className="MuiDivider-root-188"
        style={
          Object {
            "marginTop": 20,
          }
        }
      />
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
      style={Object {}}
    >
      <div
        className="MuiPaper-root-99 MuiPaper-elevation1-102 MuiPaper-rounded-100 MuiCard-root-98"
        id="card-1"
        onClick={[Function]}
        style={
          Object {
            "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.2)",
            "cursor": "pointer",
            "marginTop": 36,
          }
        }
      >
        <div
          className="MuiCardHeader-root-192"
        >
          <div
            className="MuiCardHeader-avatar-193"
          >
            <div
              className="MuiAvatar-root-198 MuiAvatar-colorDefault-199"
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "green",
                  "height": 50,
                  "width": 50,
                }
              }
            >
              U
            </div>
          </div>
          <div
            className="MuiCardHeader-content-195"
          >
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiCardHeader-title-196"
            >
              user
            </span>
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-197"
            >
              13 September 2018
            </span>
          </div>
        </div>
        <div
          className="MuiCardContent-root-126"
        >
          <h2
            className="MuiTypography-root-127 MuiTypography-headline-132 MuiTypography-gutterBottom-144"
          >
            abc
          </h2>
          <p
            className="MuiTypography-root-127 MuiTypography-body1-136"
          >
            test
          </p>
        </div>
      </div>
      <div
        className="MuiPaper-root-99 MuiPaper-elevation1-102 MuiPaper-rounded-100 MuiCard-root-98"
        id="card-2"
        onClick={[Function]}
        style={
          Object {
            "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.2)",
            "cursor": "pointer",
            "marginTop": 36,
          }
        }
      >
        <div
          className="MuiCardHeader-root-192"
        >
          <div
            className="MuiCardHeader-avatar-193"
          >
            <div
              className="MuiAvatar-root-198 MuiAvatar-colorDefault-199"
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "green",
                  "height": 50,
                  "width": 50,
                }
              }
            >
              U
            </div>
          </div>
          <div
            className="MuiCardHeader-content-195"
          >
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiCardHeader-title-196"
            >
              user
            </span>
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-197"
            >
              13 September 2018
            </span>
          </div>
        </div>
        <div
          className="MuiCardContent-root-126"
        >
          <h2
            className="MuiTypography-root-127 MuiTypography-headline-132 MuiTypography-gutterBottom-144"
          >
            def
          </h2>
          <p
            className="MuiTypography-root-127 MuiTypography-body1-136"
          >
            tester
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
</div>
`;

exports[`Profile renders Profile without description 1`] = `
<div
  className="MuiGrid-container-1"
  style={
    Object {
      "marginTop": "2%",
      "paddingBottom": "5%",
    }
  }
>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
      style={
        Object {
          "flexDirection": "column",
        }
      }
    >
      <div>
        <div
          className="MuiPaper-root-99 MuiPaper-elevation1-102 MuiPaper-rounded-100 MuiCard-root-98"
          style={
            Object {
              "boxShadow": "0px 0px 0px 0px rgba(0, 0, 0, 0.2)",
              "display": "flex",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "flex": 4,
                "flexDirection": "column",
              }
            }
          >
            <div
              className="MuiCardContent-root-126"
              style={
                Object {
                  "flex": "1 0 auto",
                }
              }
            >
              <h1
                className="MuiTypography-root-127 MuiTypography-display1-131"
              >
                user
              </h1>
              <p
                className="MuiTypography-root-127 MuiTypography-body1-136"
                style={
                  Object {
                    "marginTop": 4,
                  }
                }
              >
                
              </p>
              <p
                className="MuiTypography-root-127 MuiTypography-body1-136"
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginTop": 12,
                  }
                }
              >
                Hexanews member since Invalid date
              </p>
            </div>
          </div>
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <img
              alt="avatar"
              src={undefined}
              style={
                Object {
                  "borderRadius": "50%",
                  "height": 100,
                  "width": 100,
                }
              }
            />
          </div>
        </div>
      </div>
      <button
        className="MuiButtonBase-root-178 MuiButton-root-152 MuiButton-text-154 MuiButton-outlined-160 MuiButton-sizeSmall-175"
        disabled={false}
        id="accountButton"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "marginLeft": 20,
          }
        }
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-153"
        >
          My Account
        </span>
        <span
          className="MuiTouchRipple-root-181"
        />
      </button>
      <hr
        className="MuiDivider-root-188"
        style={
          Object {
            "marginTop": 20,
          }
        }
      />
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
      style={Object {}}
    >
      <div
        className="MuiPaper-root-99 MuiPaper-elevation1-102 MuiPaper-rounded-100 MuiCard-root-98"
        id="card-1"
        onClick={[Function]}
        style={
          Object {
            "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.2)",
            "cursor": "pointer",
            "marginTop": 36,
          }
        }
      >
        <div
          className="MuiCardHeader-root-192"
        >
          <div
            className="MuiCardHeader-avatar-193"
          >
            <div
              className="MuiAvatar-root-198 MuiAvatar-colorDefault-199"
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "green",
                  "height": 50,
                  "width": 50,
                }
              }
            >
              U
            </div>
          </div>
          <div
            className="MuiCardHeader-content-195"
          >
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiCardHeader-title-196"
            >
              user
            </span>
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-197"
            >
              13 September 2018
            </span>
          </div>
        </div>
        <div
          className="MuiCardContent-root-126"
        >
          <h2
            className="MuiTypography-root-127 MuiTypography-headline-132 MuiTypography-gutterBottom-144"
          >
            abc
          </h2>
          <p
            className="MuiTypography-root-127 MuiTypography-body1-136"
          >
            test
          </p>
        </div>
      </div>
      <div
        className="MuiPaper-root-99 MuiPaper-elevation1-102 MuiPaper-rounded-100 MuiCard-root-98"
        id="card-2"
        onClick={[Function]}
        style={
          Object {
            "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.2)",
            "cursor": "pointer",
            "marginTop": 36,
          }
        }
      >
        <div
          className="MuiCardHeader-root-192"
        >
          <div
            className="MuiCardHeader-avatar-193"
          >
            <div
              className="MuiAvatar-root-198 MuiAvatar-colorDefault-199"
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "green",
                  "height": 50,
                  "width": 50,
                }
              }
            >
              U
            </div>
          </div>
          <div
            className="MuiCardHeader-content-195"
          >
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiCardHeader-title-196"
            >
              user
            </span>
            <span
              className="MuiTypography-root-127 MuiTypography-body2-135 MuiTypography-colorTextSecondary-150 MuiCardHeader-subheader-197"
            >
              13 September 2018
            </span>
          </div>
        </div>
        <div
          className="MuiCardContent-root-126"
        >
          <h2
            className="MuiTypography-root-127 MuiTypography-headline-132 MuiTypography-gutterBottom-144"
          >
            def
          </h2>
          <p
            className="MuiTypography-root-127 MuiTypography-body1-136"
          >
            tester
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
</div>
`;
